spec/jobs/send_gpo_code_expiration_notices_job_spec.rb

```diff
@@ -0,0 +1,242 @@
+      profile.activate
+    end
+  end
+
+  let(:deactivation_reasons) do
+    %i[
+      password_reset
+      encryption_error
+      verification_cancelled
+      gpo_verification_pending_NO_LONGER_USED
+      in_person_verification_pending_NO_LONGER_USED
+    ]
+  end
+
+  let!(:users_with_profiles_with_deactivation_reasons) do
+    deactivation_reasons.map do |reason|
+      create(
+        :user,
+        :with_pending_gpo_profile,
+        code_sent_at: expired_and_notifiable_timestamp,
+      ).tap do |user|
+        profile = user.gpo_verification_pending_profile
+        profile.deactivate(reason)
+      end
+    end
+  end
+
+  let!(:user_who_started_gpo_but_then_verified_a_different_way) do
+    create(
+      :user,
+      :with_pending_gpo_profile,
+      code_sent_at: expired_and_notifiable_timestamp,
+    ).tap do |user|
+      create(:profile, :active, :verified, :with_pii, user: user)
+    end
+  end
+
+  let!(:user_who_has_an_unexpired_code) do
+    create(:user, :with_pending_gpo_profile, code_sent_at: not_expired_timestamp)
+  end
+
+  let!(:user_with_code_that_expired_too_long_ago) do
+    create(:user, :with_pending_gpo_profile, code_sent_at: too_expired_to_notify_timestamp)
+  end
+
+  before do
+    allow(IdentityConfig.store).to receive(:usps_confirmation_max_days).
+      and_return(usps_confirmation_max_days)
+  end
+
+  describe '#calculate_notification_window_bounds' do
+    it 'finds the right bounds' do
+      bounds = job.calculate_notification_window_bounds(
+        as_of: Time.zone.parse('2023-11-25 13:14:15'),
+      )
+      expected_min = Time.zone.parse('2023-10-24 00:00:00')
+      expected_max = Time.zone.parse('2023-10-26 00:00:00')
+
+      expect(bounds).to eql(expected_min..expected_max)
+    end
+  end
+
+  describe '#codes_to_send_notifications_for' do
+    def user_for(code:)
+      users = %i[
+        user_with_expired_code_who_should_be_notified
+        user_with_code_thats_not_expired_enough
+        user_with_two_expired_and_notifiable_codes
+        user_who_already_got_an_expiration_notice
+        user_who_completed_gpo
+        user_who_started_gpo_but_then_verified_a_different_way
+        user_who_has_an_unexpired_code
+        user_with_code_that_expired_too_long_ago
+      ]
+
+      users.each do |sym|
+        user = send(sym)
+        return sym if code.profile.user == user
+      end
+
+      deactivation_reasons.with_index.each do |reason, index|
+        user = users_with_profiles_in_invalid_states[index]
+        return reason if user == code.profile.user
+      end
+
+      nil
+    end
+
+    it 'returns correct codes requiring notification' do
+      expect(job.codes_to_send_notifications_for.map { |code| user_for(code: code) }).to eql(
+        [
+          :user_with_expired_code_who_should_be_notified,
+          :user_with_two_expired_and_notifiable_codes,
+        ],
+      )
+
+      expect(job.codes_to_send_notifications_for.to_a).to eql(
```

Comment:
This looks like a good place for `contain_exactly` to make it independent of order.
```suggestion
      expect(job.codes_to_send_notifications_for.to_a).to contain_exactly(
```

Reply:
Good call! Addressed in [995f9a6](https://github.com/18F/identity-idp/pull/9475/commits/995f9a600c55930aabbad01be5abfc9d9fe1d7a5)